File: r-package/tests/testthat/test-gridwell.R
fixtures_dir <- normalizePath(
    file.path(testthat::test_path(), "..", "..", "..", "fixtures"),
    mustWork = TRUE
)

load_fixture <- function(rel_path) {
    readLines(file.path(fixtures_dir, rel_path), warn = FALSE) |>
        paste(collapse = "\n")
}

# ─── Parse tests ───

test_that("gw_parse_ir works with valid JSON", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    expect_true(is(tbl, "externalptr"))
})

test_that("gw_parse_ir errors on invalid JSON", {
    expect_error(gw_parse_ir("{ not valid }"), "parse|Parse")
})

# ─── Validate tests ───

test_that("gw_validate returns empty for valid table", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    errors <- gw_validate(tbl)
    expect_equal(length(errors), 0)
})

test_that("gw_validate returns errors for invalid table", {
    json <- load_fixture("invalid/col_count_mismatch.json")
    tbl <- gw_parse_ir(json)
    errors <- gw_validate(tbl)
    expect_gt(length(errors), 0)
})

# ─── Serialize test ───

test_that("gw_to_json round-trips", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    output <- gw_to_json(tbl)
    parsed <- jsonlite::fromJSON(output, simplifyVector = FALSE)
    expect_equal(parsed$ir_version, "1.0")
})

# ─── Text render tests ───

test_that("gw_render_html produces HTML", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    html <- gw_render_html(tbl)
    expect_true(grepl("<table", html))
})

test_that("gw_render_latex produces LaTeX", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    latex <- gw_render_latex(tbl)
    expect_true(grepl("\\\\begin\\{", latex))
})

test_that("gw_render_typst produces Typst", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    typst <- gw_render_typst(tbl)
    expect_true(nchar(typst) > 0)
})

test_that("gw_render_rtf produces RTF", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    rtf <- gw_render_rtf(tbl)
    expect_true(grepl("^\\{\\\\rtf1", rtf))
})

test_that("gw_render_svg produces SVG", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    svg <- gw_render_svg(tbl)
    expect_true(grepl("<svg", svg))
})

test_that("gw_render_ansi produces output", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    ansi <- gw_render_ansi(tbl)
    expect_true(nchar(ansi) > 0)
})

test_that("gw_render_pandoc produces output", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    pandoc <- gw_render_pandoc(tbl)
    expect_true(nchar(pandoc) > 0)
})

test_that("gw_render_quarto produces output", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    quarto <- gw_render_quarto(tbl)
    expect_true(nchar(quarto) > 0)
})

test_that("gw_render dispatches by format name", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    html <- gw_render(tbl, "html")
    expect_true(grepl("<table", html))
})

test_that("gw_render errors on unknown format", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    expect_error(gw_render(tbl, "nope"), "Unknown format")
})

# ─── Binary render tests ───

test_that("gw_render_docx produces ZIP bytes", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    data <- gw_render_docx(tbl)
    expect_true(is.raw(data))
    expect_equal(data[1:4], charToRaw("PK\x03\x04"))
})

test_that("gw_render_xlsx produces ZIP bytes", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    data <- gw_render_xlsx(tbl)
    expect_true(is.raw(data))
    expect_equal(data[1:4], charToRaw("PK\x03\x04"))
})

test_that("gw_render_pptx produces ZIP bytes", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    data <- gw_render_pptx(tbl)
    expect_true(is.raw(data))
    expect_equal(data[1:4], charToRaw("PK\x03\x04"))
})

test_that("gw_render_binary dispatches by format", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    data <- gw_render_binary(tbl, "docx")
    expect_true(is.raw(data))
    expect_equal(data[1:4], charToRaw("PK\x03\x04"))
})

test_that("gw_render_binary errors on unknown format", {
    json <- load_fixture("minimal/minimal_1x1.json")
    tbl <- gw_parse_ir(json)
    expect_error(gw_render_binary(tbl, "pdf"), "Unknown binary format")
})

File: tests/test_gridwell.py
import json
from pathlib import Path

import gridwell

FIXTURES = Path(__file__).parent.parent / "fixtures"


def load_fixture(rel_path: str) -> str:
    return (FIXTURES / rel_path).read_text()


# ─── Parse tests ───


def test_parse_from_json():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    assert repr(table) == "Table(ir_version='1.0')"


def test_parse_from_dict():
    json_str = load_fixture("minimal/minimal_1x1.json")
    d = json.loads(json_str)
    table = gridwell.Table.from_dict(d)
    assert repr(table) == "Table(ir_version='1.0')"


def test_parse_ir_function():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.parse_ir(json_str)
    assert repr(table) == "Table(ir_version='1.0')"


def test_parse_invalid_json():
    try:
        gridwell.Table.from_json("{ not valid }")
        assert False, "Should have raised"
    except ValueError as e:
        assert "parse" in str(e).lower() or "JSON" in str(e)


# ─── Validate tests ───


def test_validate_valid():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    errors = table.validate()
    assert errors == []


def test_validate_invalid():
    json_str = load_fixture("invalid/col_count_mismatch.json")
    table = gridwell.Table.from_json(json_str)
    errors = table.validate()
    assert len(errors) > 0


# ─── to_json round-trip ───


def test_to_json():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    output = table.to_json()
    # Should be valid JSON
    parsed = json.loads(output)
    assert parsed["ir_version"] == "1.0"


# ─── Text render tests ───


def test_render_html():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    html = table.render_html()
    assert "<table" in html


def test_render_latex():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    latex = table.render_latex()
    assert "\\begin{" in latex


def test_render_typst():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    typst = table.render_typst()
    assert "#table(" in typst or "table(" in typst


def test_render_rtf():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    rtf = table.render_rtf()
    assert rtf.startswith("{\\rtf1")


def test_render_svg():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    svg = table.render_svg()
    assert "<svg" in svg


def test_render_ansi():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    ansi = table.render_ansi()
    assert len(ansi) > 0


def test_render_pandoc():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    pandoc = table.render_pandoc()
    assert len(pandoc) > 0


def test_render_quarto():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    quarto = table.render_quarto()
    assert len(quarto) > 0


def test_render_by_name():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    html = table.render("html")
    assert "<table" in html


def test_render_unknown_format():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    try:
        table.render("nope")
        assert False, "Should have raised"
    except ValueError as e:
        assert "nope" in str(e)


# ─── Binary render tests ───


def test_render_docx():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    data = table.render_docx()
    assert isinstance(data, bytes)
    assert data[:4] == b"PK\x03\x04"  # ZIP magic


def test_render_xlsx():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    data = table.render_xlsx()
    assert isinstance(data, bytes)
    assert data[:4] == b"PK\x03\x04"


def test_render_pptx():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    data = table.render_pptx()
    assert isinstance(data, bytes)
    assert data[:4] == b"PK\x03\x04"


def test_render_binary_by_name():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    data = table.render_binary("docx")
    assert isinstance(data, bytes)
    assert data[:4] == b"PK\x03\x04"


def test_render_binary_unknown():
    json_str = load_fixture("minimal/minimal_1x1.json")
    table = gridwell.Table.from_json(json_str)
    try:
        table.render_binary("pdf")
        assert False, "Should have raised"
    except ValueError as e:
        assert "pdf" in str(e)
